_source/_posts/2021-10-27-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,551 @@
+---
+layout: blog_post
+title: "How to Secure Your Kubernetes Cluster with OpenID Connect and RBAC"
+author: deepu-sasidharan
+by: advocate
+communities: [kubernetes, devops, security]
+description: "Learn how to setup Okta as an OpenID Connect provider using Terraform for access to Kubernetes API server."
+tags: [kubernetes, devops, oidc, okta, terraform]
+tweets:
+  - "Use #OIDC to access #Kubernetes API server securely. Set it up using Terraform."
+  - "Access #Kubernetes API server using your favorite #OIDC provider!"
+type: conversion
+image: blog/k8s-api-server-oidc/kube-login-oidc.jpg
+---
+
+A Kubernetes (k8s) cluster comprises worker machines called nodes and a control plane consisting of the API server, scheduler, etcd, controller manager, and cloud controller manager — in case of a PaaS. The containers deployed to the cluster will run in pods on the worker nodes. At the same time, the control plane takes care of scheduling, responding to requests, and managing the cluster.
+
+{% img blog/k8s-api-server-oidc/kube-architecture.jpg alt:"K8s architecture" width:"800" %}{: .center-image }
+
+When you are communicating with a Kubernetes cluster, let's say using kubectl or a client library or a tool like [KDash](https://kdash.cli.rs/), you are primarily interacting with the [Kubernetes API server](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/). The API server is responsible for managing the cluster and is responsible for handling requests from a client.
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+# Why is it required to secure the API server?
+
+The API server has multiple layers of security. By default, all communication with the API server uses TLS. Authentication is done using service account tokens, bearer tokens, basic authentication, a proxy, or client certificates, depending on the platform. In the case of PaaS like Amazon EKS, AKS, and GKE, it can also be done using custom authentication mechanisms. Once a request is authenticated, the API server can use one of several authorization mechanisms, like Attribute-based access control (ABAC) or Role-based access control (RBAC), to control access to resources. And finally, there are also admission control modules that can be configured to control resource modifications.
+
+Since the API server is the only part of the Kubernetes cluster that a client can access, it is essential to secure the API server. Unauthorized access of the API server can lead to the hijack of the entire cluster and maybe even your infrastructure, data theft, and so on. Hence, configuring users and roles properly is a must to secure the cluster, especially i9n organizations where more than one user can access the cluster.
```

Comment:
```suggestion
Since the API server is the only part of the Kubernetes cluster that a client can access, it is essential to secure the API server. Unauthorized access of the API server can lead to hijacking of the entire cluster and maybe even your infrastructure, data theft, and so on. Hence, configuring users and roles properly is a must to secure the cluster, especially i9n organizations where more than one user can access the cluster.
```